File: src/datatypes/sensor.ts
import { BaseData, CameraType, ConnectionType } from './types';

// Motion sensors
export class AccelerometerData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    x!: number;
    y!: number;
    z!: number;
    accuracy?: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<AccelerometerData>) {
        Object.assign(this, data);
    }
}

export class GyroscopeData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    x!: number;
    y!: number;
    z!: number;
    accuracy?: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<GyroscopeData>) {
        Object.assign(this, data);
    }
}

export class MagnetometerData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    x!: number;
    y!: number;
    z!: number;
    accuracy?: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<MagnetometerData>) {
        Object.assign(this, data);
    }
}

// Location sensors
export class GPSData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    latitude!: number;
    longitude!: number;
    altitude?: number;
    accuracy?: number;
    speed?: number;
    bearing?: number;
    satellites?: number;
    provider?: string;
    metadata?: Record<string, any>;

    constructor(data: Partial<GPSData>) {
        Object.assign(this, data);
    }
}

// Health sensors
export class HeartRateData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    bpm!: number;
    confidence?: number;
    rr_intervals?: number[];
    metadata?: Record<string, any>;

    constructor(data: Partial<HeartRateData>) {
        Object.assign(this, data);
    }
}

export class ECGData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    voltage!: number[];
    time!: number[];
    rhythm_classification?: string;
    heart_rate?: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<ECGData>) {
        Object.assign(this, data);
    }
}

export class BloodOxygenData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    spo2!: number;
    confidence?: number;
    raw_values?: number[];
    metadata?: Record<string, any>;

    constructor(data: Partial<BloodOxygenData>) {
        Object.assign(this, data);
    }
}

export class StressData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    stress_score!: number;
    stress_level?: string;
    hrv?: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<StressData>) {
        Object.assign(this, data);
    }
}

// Environmental sensors
export class ProximityData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    distance!: number;
    near!: boolean;
    metadata?: Record<string, any>;

    constructor(data: Partial<ProximityData>) {
        Object.assign(this, data);
    }
}

export class LightData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    lux!: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<LightData>) {
        Object.assign(this, data);
    }
}

export class PressureData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    hectopascals!: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<PressureData>) {
        Object.assign(this, data);
    }
}

export class TemperatureData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    celsius!: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<TemperatureData>) {
        Object.assign(this, data);
    }
}

export class HumidityData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    percentage!: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<HumidityData>) {
        Object.assign(this, data);
    }
}

// Activity sensors
export class StepCountData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    steps!: number;
    activity_type?: string;
    confidence?: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<StepCountData>) {
        Object.assign(this, data);
    }
}

// Audio sensors
export class AudioLevelData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    db!: number;
    peak_db?: number;
    volume?: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<AudioLevelData>) {
        Object.assign(this, data);
    }
}

// System sensors
export class BatteryData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    percentage!: number;
    charging!: boolean;
    power_source?: string;
    temperature?: number;
    voltage?: number;
    current?: number;
    metadata?: Record<string, any>;

    constructor(data: Partial<BatteryData>) {
        Object.assign(this, data);
    }
}

export class NetworkData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    type!: ConnectionType;
    state?: string;
    strength?: number;
    carrier?: string;
    roaming?: boolean;
    cellular_technology?: string;
    is_metered?: boolean;
    dns_servers?: string[];
    gateway?: string;
    metadata?: Record<string, any>;

    constructor(data: Partial<NetworkData>) {
        Object.assign(this, data);
    }
}

export class ScreenStateData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    screen_on!: boolean;
    brightness?: number;
    orientation?: string;
    metadata?: Record<string, any>;

    constructor(data: Partial<ScreenStateData>) {
        Object.assign(this, data);
    }
}

// Camera and vision sensors
export class CameraData implements BaseData {
    timestamp!: Date;
    device_id!: string;
    camera_type!: CameraType;
    light_level?: number;
    scene_type?: string;
    objects?: Record<string, any>;
    face_detection?: Record<string, any>;
    focus_distance?: number;
    flash_state?: string;
    zoom_level?: number;
    capture_mode?: string;
    metadata?: Record<string, any>;

    constructor(data: Partial<CameraData>) {
        Object.assign(this, data);
    }
} 

File: src-tauri/src/datatypes/sensor.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use crate::datatypes::types::{BaseData, CameraType, ConnectionType, Metadata};

// Motion sensors
#[derive(Debug, Serialize, Deserialize)]
pub struct AccelerometerData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub x: f32,
    pub y: f32,
    pub z: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accuracy: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GyroscopeData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub x: f32,
    pub y: f32,
    pub z: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accuracy: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct MagnetometerData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub x: f32,
    pub y: f32,
    pub z: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accuracy: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

// Location sensors
#[derive(Debug, Serialize, Deserialize)]
pub struct GPSData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub latitude: f64,
    pub longitude: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub altitude: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accuracy: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub speed: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bearing: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub satellites: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

// Health sensors
#[derive(Debug, Serialize, Deserialize)]
pub struct HeartRateData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub bpm: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub confidence: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rr_intervals: Option<Vec<f32>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ECGData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub voltage: Vec<f32>,
    pub time: Vec<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rhythm_classification: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub heart_rate: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BloodOxygenData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub spo2: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub confidence: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub raw_values: Option<Vec<f32>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct StressData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub stress_score: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stress_level: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hrv: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

// Environmental sensors
#[derive(Debug, Serialize, Deserialize)]
pub struct ProximityData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub distance: f32,
    pub near: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LightData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub lux: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PressureData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub hectopascals: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TemperatureData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub celsius: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct HumidityData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub percentage: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

// Activity sensors
#[derive(Debug, Serialize, Deserialize)]
pub struct StepCountData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub steps: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub activity_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub confidence: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

// Audio sensors
#[derive(Debug, Serialize, Deserialize)]
pub struct AudioLevelData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub db: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub peak_db: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub volume: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

// System sensors
#[derive(Debug, Serialize, Deserialize)]
pub struct BatteryData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub percentage: i32,
    pub charging: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub power_source: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub voltage: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct NetworkData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub connection_type: ConnectionType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub strength: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub carrier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub roaming: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cellular_technology: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_metered: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dns_servers: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gateway: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ScreenStateData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub screen_on: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub brightness: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub orientation: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

// Camera and vision sensors
#[derive(Debug, Serialize, Deserialize)]
pub struct CameraData {
    pub timestamp: DateTime<Utc>,
    pub device_id: String,
    pub camera_type: CameraType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub light_level: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scene_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub objects: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub face_detection: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub focus_distance: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub flash_state: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub zoom_level: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub capture_mode: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
}

// Implement JSON conversion for all sensor types
macro_rules! impl_json_conversion {
    ($($t:ty),*) => {
        $(
            impl $t {
                pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
                    serde_json::from_str(json)
                }

                pub fn to_json(&self) -> Result<String, serde_json::Error> {
                    serde_json::to_string(self)
                }
            }
        )*
    };
}

impl_json_conversion!(
    AccelerometerData, GyroscopeData, MagnetometerData,
    GPSData, HeartRateData, ECGData, BloodOxygenData,
    StressData, ProximityData, LightData, PressureData,
    TemperatureData, HumidityData, StepCountData,
    AudioLevelData, BatteryData, NetworkData,
    ScreenStateData, CameraData
); 